edgraph/server.go

```diff
@@ -1294,6 +1308,17 @@ func (s *Server) doQuery(ctx context.Context, req *Request) (resp *api.Response,
 		ostats.Record(ctx, x.NumMutations.M(1))
 	}
 
+	if isQuery || isMutation {
+		m := req.req.Query
+		if !isQuery {
+			m = req.req.String()
+		}
+		if strings.Contains(m, "#") {
```

Comment:
This is not a very robust logic and this should be implemented at the time of parsing the query instead. I am also curious how one would specify this tag in a client.

Reply:
We have a very narrow window(to have it in beta) and we need this. Any improvement can come later. We can open a Jira Ticket to follow up.

I gave just the idea and I couldn't wait anyone to pick it up(feels like nobody would pick it). if you have a better way you can add your changes now or later. But this will help us a LOT during debugging in the cloud.

Reply:
Any client will work the same. It is just a comment, which is "native" for the queries and mutations.

Reply:
Let's talk about this. I understand the urgency of it but I would really like to make sure the quality of release to the best that we can.

Reply:
To me the two important things are:

 1) Is this a good, long-term way to tag queries that will work for the users?
 2) is this a safe change?

How it is coded is less critical, because if the user behavior is desirable, and it works, we can easily move it to a parsing layer (or somewhere else) later.

Similarly for testing - for now we can test manually, and next release we can add those same checks as automated unit test.

So the key as I see it is the behavior that is exposed to the users:  the first ```#tag:<sometext>``` comment in a query will cause a new attribute tag=<sometext> to be added to the otrace span. This seems like a good behavior to me, but this is the key item I think we should discuss.

Is it safe? Since the new code paths are "protected" first by a check if '#' is in the query, and then a regex match, it seems that this is very unlikely to break existing queries. It's also a little safer because the only behavior is adding an otrace attribute, vs changing any query behavior.

Reply:
I agree with most of what you said, Damon. These are my concerns.

I can't see right now that this would be a robust change. I am wondering what if my query itself contains the expected tag. It could show up as part of a string in the query. Will our code still work?

Another question I have is, will this be extensible to our dgraph clients, what about the JSON API for querying data? What about mutations and upserts later? The language change has to be well defined too and that would be true when lexer and parser can make sense of it correctly. 

We would need tests to ensure that we don't break any existing behaviour, I can't think of any other way to ensure that.

Why don't we use our TXNs API instead of modifying the query language? We could add another parameter to the API and let the user pass the tag there instead.

Reply:
@damonfeldman I'd like to echo Aman's concerns above.  The way this change is written right now is not very robust.  Some fundamental issues are:
- If the query itself contains the tag (not in a comment, but inside the query) then this will probably get picked up by the regex, and this is certainly undesirable behavior and may cause panics / errors since the user may not have tracing set up
- Furthermore this mixes up the logic of parsing the query (which belongs in a different package) into the `edgraph` package here which should not contain this kind of logic.  We have a lexer, parser, etc. for this purpose and should not be matching regexes due to the potential issues pointed out above.

I agree with @mangalaman93 above that we could modify the transaction API to let the user pass the tags there instead.  There are some subtleties there to think about but I think that is a much cleaner way to make such a change.

Reply:
I hope you read this. I'm addressing every point mentioned above. I agree that the code could be improved and put in a place (if there is a place for it) to look neat and tidy. However, all concerns raised are invalid. Which leads me to believe that you didn't test your theories before raising them. You didn't tested out the code. Mentally or in practice.

----------------

The point about being "robust" can be assessed later if your concern is beauty or performance - We need as it is now to be able to obtain data in the surgical search for problems with a specific query. 

The code only checks if it has a comment which starts with #. If there is no hashtag first it will be ignored. Yes, if there is another field with # it will be positive. But there's another check, with the regex that looks for a more complex pattern. If the pattern does not match. It doesn't matter if there are other hashtags in other corners of the query it will be ignored. That logic is simple to read from the code.

>I am wondering what if my query itself contains the expected tag. It could show up as part of a string in the query. Will our code still work?

I'm not sure I understood the question well. Because the answer is obvious, see, If your query has a comment (the comment is native to Dgraph, we support comments in queries and mutations, you can have 90% of the query only comments and it won't be a problem) with the expected tag. it will just append the tag to the span trace that was already created during the start of the transaction. It's just an append.

The part "It could show up as part of a string in the query" I don't understand. Do you mean the response or the body of the trace? Yes, all comments whether tagged or not - with this PR or not - will be present in the trace. This is the default behavior, I didn't define it with this code. But it would never be part of the response. That's impossible. 

>Another question I have is, will this be extensible to our dgraph clients, what about the JSON API for querying data? What about mutations and upserts later?

For those who understand the flow of the query execution, the answer to this question would be simple. The parsing of the query happens way before this step in which I put this check. The query transformation has already been done. It doesn't matter if the query will come in JSON or DQL or RDF. It was already parsed well before reaching this step of the code. 

The only thing missing is mutation in JSON. Since JSON does not support comments. Mutations in JSON will be stopped long before the transaction even starts. Because JSON does not support comments.

Yes. This code works for RDF and Upserts (Read the comment of the PR - I have put an example for RDF). In the case of Upserts they only work if the comment is in the query block, if you put it in the mutation block it will be ignored.

> The language change has to be well defined too and that would be true when lexer and parser can make sense of it correctly.

There's NO language change here. I'm using an existing feature which is comments.

> Why don't we use our TXNs API instead of modifying the query language?

Again, there's NO language change... How can't you see that? oO

> We could add another parameter to the API and let the user pass the tag there instead.

That would be nice, if not take months or years to do. I mean, pretty sure it would be fast to code. But the problem is "willing" and collaboration.

Reply:
@joshua-goldstein 

>If the query itself contains the tag (not in a comment, but inside the query) then this will probably get picked up by the regex

Never. We can bet. :P I can bet 5 months salary on this.

First: The query has to have the hashtag. Which is possible. You can have a predicate like <#.hashtag> or a value that has hashtags. That is fine. But only the first check it will grab it.

Second: The regex expression  `(?m)#tag:(\w+)` is used to match a pattern that begins with the string "#tag:" followed by one or more word characters without space. If your query doesn't start with # and doesn't have the word "tag" it will be ignored. You can test out in https://regex101.com/r/SB97n7/1 if you wanna confirm. In this example we have two valig tags that the regex will capture. BUT the lines 557 and 559 in the code I'm selecting only the first found with `querytag[1]`.

So the Regex will trap only one match and that match has to start with #tag: and have a word after the colon. If you don't do like this, it will be ignored.

>Furthermore this mixes up the logic of parsing the query

Not really. I'm not parsing anything. Just checking if it exist and picking. Not transforming anything. It is similar to parsing, but I'm not transforming the query into an object that will be used in another part of the code. That's the meaning of parse. Transform something meaningful for your architecture.


>I agree with @mangalaman93 above that we could modify the transaction API to let the user pass the tags there instead. There are some subtleties there to think about but I think that is a much cleaner way to make such a change.

I also agree, but I'm pretty sure this code is fine. And you can do whatever you want later. All concerns above are not valid. And we need this feature for customer success and other debug activities. We don't have time to wait months for someone to make the decision to do it. I commented on this feature in Slack weeks ago. And it was ignored, nobody wanted to work or discuss it. I went there and did it. Well, cuz it's simple and easy. Now, if you want to do something more bureaucratic and supposedly better and safer. I support it and I'm glad this caught your attention and you're interested in improving. That is good.

Reply:
By the way, I would be very happy to review any designs you guys come up with. Truly glad you guys invested your time developing something on this idea.

The idea of adding a new entry in the API is good, but it seems to me that you want to use a Bazooka against an ant. Create an entire design for a single tag. Like, spending a fortune on time for such small thing.

Like, there's a (fake) joke about Russians and Americans in space. Where the US spent billions of dollars to develop a pen that could write in space. While the Russians used a simple pencil.

Reply:
These are all good points. I do think the change is pretty safe, since annotations are lower impact than query or process changes, but even so, testing is always important.

Let's do what I think both Aman and Michel are suggesting and set up a design meeting on this once the v23 crunch is over. We'll end up with an ideal solution in the end, I am sure.